### ISSUE TITLE
GUEST_CHECK_PADDR width issues

### ISSUE THREAD

**Issue by djwillia** [Solo5/solo5#117]
On the host side, `ukvm` checks addresses passed into the guest to make sure that they don't exceed the guest memory.  The macro `GUEST_CHECK_PADDR` (https://github.com/Solo5/solo5/blob/master/ukvm/ukvm-private.h#L68) is supposed to do this.

There was a question about whether uint64_t pointers should be checked with `uaddll_overflow` or `uaddl_overflow`.   It seems that this is not entirely straightforward, as different compilers may have different builtins or macros to do this that treat `unsigned long` and `unsigned long long` differently. 

The start of the discussion between @mato and @djwillia was in this PR (https://github.com/Solo5/solo5/pull/116)

**Comment by mato** [comment]
@djwillia Can you check if the version in #124 works for you without any compiler warnings on clang/osx? (But don't merge yet)

I tested with:
- Debian stable, GCC 4.9.2, uses macros
- Debian stretch, GCC 6.2.0, uses builtins
- Debian stretch, clang 3.8.1, uses builtins (manually set `UKVM_CC=clang` appropriately)
- Travis, GCC 4.something, uses macros

**Comment by mato** [comment]
(commenting here rather than in the PR(s))

@djwillia Regarding the issue with `size_t` and using `%zu` for it, that's probably fine, however I have yet to find time to re-run my tests on the various compilers with it.

What I'm more worried about is that the change relies on using the generic `__builtin_add_overflow` rather than the more strongly typed `__builtin_uaddl_overflow`. Based on the [clang](http://clang.llvm.org/docs/LanguageExtensions.html#checked-arithmetic-builtins) and [gcc](https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html) documentation this _seems_ fine, but I'm not 100% sure.

**Comment by mato** [comment]
I've updated #124 and manually tested it with the following test program:
````
#include <stdint.h>
#include <stddef.h>

#include "ukvm/ukvm-private.h"
#include <assert.h>

#define TEST_CHECK_PADDR(p, l, sz) \
    {                                                                          \
        uint64_t __e;                                                          \
        if ((p >= l) || add_overflow(p, sz, __e) || (__e >= l))                \
            warnx("%s:%d: Invalid guest access: "                              \
                    "paddr=0x%" PRIx64 ", sz=%zu",                             \
                    __FILE__, __LINE__, p, sz);                                \
    }

int main(int argc, char* argv[])
{
    volatile uint64_t paddr;
    volatile uint64_t l = 0x2000000;
    volatile size_t sz;

    paddr = 0x0;
    sz = 0x1000;
    TEST_CHECK_PADDR(paddr, l, sz);

    paddr = 0x1ffffff;
    TEST_CHECK_PADDR(paddr, l, sz);

    sz = -1;
    TEST_CHECK_PADDR(paddr, l, sz);

    paddr = 0x0;
    TEST_CHECK_PADDR(paddr, l, sz);

    paddr = -0x1000;
    sz = 0;
    TEST_CHECK_PADDR(paddr, l, sz);

    return 0;
}
````
This produces the expected, identical, result on all the Debian compilers above:
````
test-overflow: test-overflow.c:27: Invalid guest access: paddr=0x1ffffff, sz=4096
test-overflow: test-overflow.c:30: Invalid guest access: paddr=0x1ffffff, sz=18446744073709551615
test-overflow: test-overflow.c:33: Invalid guest access: paddr=0x0, sz=18446744073709551615
test-overflow: test-overflow.c:37: Invalid guest access: paddr=0xfffffffffffff000, sz=0
````
I've also verified the assembly code generated by the compilers (hence the use of `volatile` in the test, otherwise most of it just gets optimized out) and the compiler is using the correct additions.

This looks good to go.

**Comment by mato** [comment]
Fixed in #124.
